third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -182,13 +185,42 @@ where
             loader,
             ty_depth_checker,
             layout_converter,
+            ref_state: RefCheckState::new(),
         };
 
         let function = Rc::new(function);
         // TODO: remove Self::paranoid_type_checks fully to be replaced
         // with the static RuntimeTypeCheck trait
-        if interpreter.vm_config.paranoid_type_checks {
-            interpreter.dispatch_execute_main::<FullRuntimeTypeCheck>(
+        // Note: we have organized the code below from most-likely config to least-likely config.
+        if interpreter.vm_config.paranoid_type_checks && !interpreter.vm_config.paranoid_ref_checks
+        {
+            interpreter.dispatch_execute_main::<FullRuntimeTypeCheck, NoRuntimeRefCheck>(
+                data_cache,
+                resource_resolver,
+                module_storage,
+                gas_meter,
+                traversal_context,
+                extensions,
+                function,
+                args,
+            )
+        } else if interpreter.vm_config.paranoid_type_checks
+            && interpreter.vm_config.paranoid_ref_checks
+        {
+            interpreter.dispatch_execute_main::<FullRuntimeTypeCheck, FullRuntimeRefCheck>(
+                data_cache,
+                resource_resolver,
+                module_storage,
+                gas_meter,
+                traversal_context,
+                extensions,
+                function,
+                args,
+            )
```

Comment:
IMO it's a bit concerning here that we're duplicating the call 4 times, just for different type arguments.

With a bit of macro we can at least share the other arguments:
- https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=56b23d7cfe61c3f9e776bb708e1d27d5